### PR TITLE
make wallet restoration partially multi-core

### PR DESCRIPTION
**Pull request by cleverca22**
## Description
this will run the prefilter-block and getblock parts of wallet restore in parallel, across multiple cores, and buffer ahead several batches

then allow the existing acid-state updates to happen in a single-thread, on a stream of prefetch prefiltered blocks

it reduces the estimated wallet restore time from 2h to ~11-30minutes

a util to measure wallet restore time should be considered, so these measurements can be made more reproducible

```
[clever@amd-nixos:~/dedup/staging]$ curl -s --cert tls/client/client.pem --cacert tls/server/ca.crt https://127.0.0.1:43821/api/v1/wallets | jq .data[1].syncState.data.throughput
{
  "quantity": 2163,
  "unit": "blocksPerSecond"
}
```
the highest speed i have seen so far when restoring

## Type of change

- [ ] 🐞 Bug fix (non-breaking change which fixes an issue)
- [ ] 🛠 New feature (non-breaking change which adds functionality)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)
- [ ] 🏭 Refactoring that does not change existing functionality but does improve things like code readability, structure etc
- [ ] 🔨 New or improved tests for existing code
- [ ] ⛑ git-flow chore (backport, hotfix, etc)

## Developer checklist

- [ ] I have read the [style guide](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/style-guide.md) document, and my code follows the code style of this project.
- [ ] If my code deals with exceptions, it follows the [guidelines](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/exceptions.md).
- [ ] I have updated any documentation accordingly, if needed. Documentation changes can be reflected in opening a PR on [cardanodocs.com](https://github.com/input-output-hk/cardanodocs.com), amending the inline [Haddock](https://www.haskell.org/haddock/) comments, any relevant README file or one of the document listed in the [docs](https://github.com/input-output-hk/cardano-sl/tree/develop/docs) directory.
- [ ] CHANGELOG entry has been added and is linked to the correct PR on GitHub.

## Testing checklist

- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

## How to merge

Send the message `bors r+` to merge this PR. For more information, see
[`docs/how-to/bors.md`](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/how-to/bors.md).

### CODE REVIEW

File: lib/src/Pos/Launcher/Mode.hs [hunk 1]
```diff
@@ -75,12 +77,17 @@ instance HasSlottingVar InitModeContext where
     slottingTimestamp = imcSlottingVar_L . _1
     slottingVar = imcSlottingVar_L . _2
 
+-- how would i embed this value inside the InitMode structure? (which i would have to turn into a data record?)
```
Comment (erikd) [inline lib/src/Pos/Launcher/Mode.hs]: `InitMode` is a typeclass so you can't.

File: db/src/Pos/DB/Epoch/Index.hs [hunk 2]
```diff
@@ -35,20 +41,45 @@ import           Pos.Core (LocalSlotIndex (..), SlotCount, localSlotIndices)
...
+      writeTVar (icCache indexCache) mempty
+      pure ()
+  atomically clearCache
+
+data IndexCache = IndexCache
+    { icCache :: TVar (Map FilePath MaybeCacheEntry)
+    , icMaxOpenFiles :: Int
+    }
+
+data MaybeCacheEntry = Opening | Opened CacheEntry
```
[the remark this answers is not in the document]
Comment (cleverca22) [reply db/src/Pos/DB/Epoch/Index.hs]: the original version had 2 fields, a `Handle` and a `Lock`
i also plan to add another field of when it was last used, so the cache expiry is smarter

File: lib/src/Pos/Launcher/Mode.hs (hunk 1 above)
Comment (cleverca22) [reply lib/src/Pos/Launcher/Mode.hs]: i think reflection and `Given` would be simpler, but i can see that being a pain to thread thru half of cardano...

Comment (cleverca22) [reply lib/src/Pos/Launcher/Mode.hs]: i have moved it into the `NodeDBs` structure, so it lives beside the paths of the db files and such

**Pull request closed without merging**